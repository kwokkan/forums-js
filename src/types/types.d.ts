
interface IForumSessionUser {
    id: number;
    name: string;
    joinedDate: number;
}

interface INextSession extends any {
    forumsUser?: IForumSessionUser;
}

interface INextProvider extends any {
    id: string;
    name: string;
}

declare module "next-auth" {
    interface INextAccount extends any {
        provider: string;
        id: string;
    }

    export interface IProviderOptions {
        providers: any[];
        callbacks?: IProviderOptionsCallback;
        events?: INextEventsOptions;
        pages?: INextPageOptions;
        debug?: boolean;
    }

    interface IProviderOptionsCallback {
        signIn?: (user: INextUser, account: INextAccount, profile: INextProfile) => Promise<bool>;
        redirect?: (url: string, baseUrl: string) => Promise<string>;
        session?: (session: INextSession, user: INextUser, sessionToken: string) => Promise<INextSession>;
        jwt?: (token: INextToken, user: INextUser, account: INextAccount, profile: INextProfile, isNewUser: boolean) => Promise<INextToken>;
    }

    interface INextEventsOptions {
        signIn?: (message: any) => Promise;
        signOut?: (message: any) => Promise;
        createUser?: (message: any) => Promise;
        linkAccount?: (message: any) => Promise;
        session?: (message: any) => Promise;
        error?: (message: any) => Promise;
    }

    interface INextPageOptions {
        signIn?: string;
        signOut?: string;
        error?: string;
        verifyRequest?: string;
        newUser?: string;
    }

    interface INextProfile extends any {
    }


    interface INextToken extends any {
        forumsUser?: IForumSessionUser;
    }

    interface INextUser extends any {
        name: string;
        forumsUser?: IForumSessionUser;
    }

    function handler(req: IncomingMessage, res: ServerResponse, options: IProviderOptions): void;

    export = handler;
}

declare module "next-auth/client" {
    export const Provider: any;

    export function getSession(context: { req: IncomingMessage }): Promise<INextSession>;

    export function providers<T>(context: GetServerSidePropsContext<T>): Promise<INextProvider[]>;

    export function signIn(provider?: string, args?: { callbackUrl?: string }): void;

    export function signOut(): void;

    export function useSession(): [INextSession | null, boolean];
}

declare module "next-auth/providers" {
    interface IProviderConfig {
        clientId: string,
        clientSecret?: string,
        scope?: string[] | string
    }

    export = {
        GitHub: (config: IProviderConfig) => any
    };
}
