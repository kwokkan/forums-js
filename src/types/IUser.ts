export interface IUser {
    id: number;
    name: string;
    joinedDate: Date;
}
