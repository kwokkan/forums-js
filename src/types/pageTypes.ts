import { GetServerSideProps } from "next";
import { ParsedUrlQuery } from "querystring";

export type GetTypedServerSideProps<TParams extends ParsedUrlQuery, TProps = { [key: string]: any }> = GetServerSideProps<TProps, TParams>;
