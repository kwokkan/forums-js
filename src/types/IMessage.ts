import { IUser } from "./IUser";

export interface IMessage {
    id: number;
    created: Date;
    content: string;
    user: IUser;
}
