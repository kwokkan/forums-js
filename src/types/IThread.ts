import { IMessage } from "./IMessage";

export interface IThread {
    id: number;
    created: number;
    name: string;
    messages: IMessage[];
}
