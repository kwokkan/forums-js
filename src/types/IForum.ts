import { IThread } from "./IThread";

export interface IForum {
    id: number;
    created: Date;
    name: string;
    description?: string;
    threads?: IThread[];
}
