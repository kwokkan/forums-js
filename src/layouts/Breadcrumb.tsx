import { Breadcrumb as RBreadcrumb, Icon } from "rsuite";
import { IBreadcrumb } from "../types/IBreadcrumb";

interface IProps {
    breadcrumbs: IBreadcrumb[];
}

export function Breadcrumb({ breadcrumbs }: IProps) {
    return (
        <RBreadcrumb>
            <RBreadcrumb.Item href="/" passHref title="Home">
                <Icon icon="home" />
            </RBreadcrumb.Item>

            {breadcrumbs.map(x =>
                <RBreadcrumb.Item key={x.url} title={x.title} href={x.url}>
                    {x.title}
                </RBreadcrumb.Item>
            )}
        </RBreadcrumb>
    );
}
