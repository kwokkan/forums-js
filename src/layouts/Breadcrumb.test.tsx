/**
 * @jest-environment jsdom
*/
import { create } from "react-test-renderer";
import { IBreadcrumb } from "../types/IBreadcrumb";
import { Breadcrumb } from "./Breadcrumb";

test("Renders with breadcrumbs", () => {
    const breadcrumbs: IBreadcrumb[] = [
        {
            title: "First",
            url: "/urls/1"
        },
        {
            title: "Second",
            url: "/urls/2"
        }
    ];

    const tree = create(
        <Breadcrumb breadcrumbs={breadcrumbs} />
    ).toJSON();

    expect(tree).toMatchSnapshot();
});
