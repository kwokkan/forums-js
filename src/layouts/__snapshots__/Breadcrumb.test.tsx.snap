// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders with breadcrumbs 1`] = `
<ol
  aria-label="breadcrumbs"
  className="rs-breadcrumb"
  role="navigation"
>
  <li
    className="rs-breadcrumb-item"
  >
    <a
      href="/"
      onClick={[Function]}
      passHref={true}
      title="Home"
    >
      <i
        className="rs-icon rs-icon-home"
      />
    </a>
  </li>
  <li
    className="rs-breadcrumb-separator"
  >
    /
  </li>
  <li
    className="rs-breadcrumb-item"
  >
    <a
      href="/urls/1"
      onClick={[Function]}
      title="First"
    >
      First
    </a>
  </li>
  <li
    className="rs-breadcrumb-separator"
  >
    /
  </li>
  <li
    className="rs-breadcrumb-item"
  >
    <a
      href="/urls/2"
      onClick={[Function]}
      title="Second"
    >
      Second
    </a>
  </li>
</ol>
`;
